compiler/rustc_mir/src/transform/abort_unwinding_calls.rs

```diff
@@ -0,0 +1,107 @@
+/// Rust-defined functions never unwind a well-behaved Rust program needs to
+/// catch unwinding from foreign functions and force them to abort.
+///
+/// This pass walks walks over all functions calls which may possibly unwind,
+/// and if any are found sets their cleanup to a block that aborts the process.
+/// This forces all unwinds, in panic=abort mode happening in foreign code, to
+/// trigger a process abort.
+#[derive(PartialEq)]
+pub struct AbortUnwindingCalls;
+
+impl<'tcx> MirPass<'tcx> for AbortUnwindingCalls {
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        // This pass only runs on functions which themselves cannot unwind,
+        // forcibly changing the body of the function to structurally provide
+        // this guarantee by aborting on an unwind. If this function can unwind,
+        // then there's nothing to do because it already should work correctly.
+        //
+        // Here we test for this function itself whether its ABI allows
+        // unwinding or not.
+        let body_flags = tcx.codegen_fn_attrs(body.source.def_id()).flags;
+        let body_ty = tcx.type_of(body.source.def_id());
+        let body_abi = match body_ty.kind() {
+            ty::FnDef(..) => body_ty.fn_sig(tcx).abi(),
+            ty::Closure(..) => Abi::RustCall,
+            ty::Generator(..) => Abi::Rust,
+            _ => span_bug!(body.span, "unexpected body ty: {:?}", body_ty),
+        };
+        if layout::fn_can_unwind(tcx, body_flags, body_abi) {
+            return;
+        }
+
+        // Look in this function body for any basic blocks which are terminated
+        // with a function call, and whose function we're calling may unwind.
+        // This will filter to functions with `extern "C-unwind"` ABIs, for
+        // example.
+        let mut calls_to_terminate = Vec::new();
+        for (id, block) in body.basic_blocks().iter_enumerated() {
+            if block.is_cleanup {
+                continue;
+            }
+
+            let (func, source_info) = match &block.terminator {
+                Some(Terminator { kind: TerminatorKind::Call { func, .. }, source_info }) => {
+                    (func, source_info)
+                }
+                _ => continue,
+            };
+            let ty = func.ty(body, tcx);
+            let sig = ty.fn_sig(tcx);
+            let flags = match ty.kind() {
+                ty::FnPtr(_) => CodegenFnAttrFlags::empty(),
+                ty::FnDef(def_id, _) => tcx.codegen_fn_attrs(*def_id).flags,
+                _ => span_bug!(source_info.span, "invalid callee of type {:?}", ty),
+            };
+            if !layout::fn_can_unwind(tcx, flags, sig.abi()) {
+                continue;
+            }
+            calls_to_terminate.push(id);
+        }
+
+        // If nothing was possibly unwinding then there's nothing to modify in
+        // this function so we can skip it.
+        if calls_to_terminate.is_empty() {
```

Comment:
I feel like I'm missing something. When we call a extern "Rust" function, for example, it will have an unwind edge, right? But if this is panic=abort, we want to remove that unwind edge, right?

Reply:
Or do we handle that at MIR build time?

Reply:
Yeah it will have an unwind edge, primarily for borrowck from what I've surmised. Previously the `no_landing_pads` pass would remove it but now this relys on codegen to simply ignore the landing pad (which it currently does today). This means that functions in the MIR will have landing pads which end up just getting ignored.

I think it makes more sense here though to adjust landing pads to simply get removed, even for functions which can't unwind that don't call anything which might unwind.